### PR TITLE
Addition of Stats Exporter for Microsoft Azure

### PR DESCRIPTION
**Pull request by jimmy-long**
This is in relation to issue #751 which I opened for my team's senior design project at NC State University.

### CODE REVIEW

File: packages/opencensus-exporter-azure/package.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "@opencensus/exporter-azure",
+  "version": "0.0.19",
```
Comment (mayurkale22) [inline packages/opencensus-exporter-azure/package.json:3]:
```suggestion
  "version": "0.0.21",
```

File: packages/opencensus-exporter-azure/src/azure-stats.ts [hunk 2]
```diff
@@ -0,0 +1,302 @@
+import {
```
Comment (mayurkale22) [inline packages/opencensus-exporter-azure/src/azure-stats.ts:1]: Please include the license header in all the newly added files.

File: packages/opencensus-exporter-azure/src/azure-trace.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+import * as ApplicationInsights from 'applicationinsights';
+import { IllegalOptionsError } from './azure-stats';
+
+export interface TraceParams {
+    registeredViews: View[];
+    registeredMeasures: Measure[];
+    recordedData: { [key: string]: AggregationData[] };
+}
+
+export interface AzureTraceExporterOptions extends ExporterConfig {
```
Comment (mayurkale22) [inline packages/opencensus-exporter-azure/src/azure-trace.ts:22]: I would suggest to keep these types in a separate file, maybe in `types.ts`?

Comment (mayurkale22) [reply packages/opencensus-exporter-azure/src/azure-trace.ts:22]: Please include the JSDoc comment.

File: packages/opencensus-exporter-azure/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,2 @@
+export * from './azure-stats';
+export * from './azure-trace';
```
Comment (mayurkale22) [inline packages/opencensus-exporter-azure/src/index.ts:2]: Usually we keep trace and stats exporters in separate packages, not sure if you want to do the same thing here.

File: packages/opencensus-exporter-azure/src/azure-stats.ts [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    registeredMeasures: Measure[];
+    recordedData: { [key: string]: AggregationData[] };
+}
+
+export enum ExportMode {
+    SINGLE_VALUE = 0,
+    BATCH = 1
+}
+
+export enum AggregationMethod {
```
Comment (mayurkale22) [inline packages/opencensus-exporter-azure/src/azure-stats.ts:35]: Same here, pls keep these types in separate file.

File: packages/opencensus-exporter-azure/src/azure-stats.ts [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+    prefix: 'OpenCensus',
+    exportMode: ExportMode.SINGLE_VALUE,
+    aggregationMethod: AggregationMethod.AVERAGE,
+    logger: logger.logger()
+}
+
+/**
+ * Formats and sends Stats to Azure Monitor.
+ */
+export class AzureStatsExporter implements StatsEventListener {
```
Comment (mayurkale22) [inline packages/opencensus-exporter-azure/src/azure-stats.ts:114]: Do you plan to add unit tests for all these functionality?

File: package.json [hunk 7]
```diff
@@ -45,5 +45,8 @@
     "lerna": "^2.11.0",
     "typedoc": "^0.17.0",
     "typescript": "^3.0.0"
+  },
+  "dependencies": {
+    "applicationinsights": "^1.7.2"
```
Comment (mayurkale22) [inline package.json:50]: I think this is not needed here.

**Comment by cdietschrun** [comment]
Hi all involved, my name's Chris Dietsch (MSFT alias chdietsc) and I was the Microsoft sponsor for this project which was a university senior design project. Today was the handoff for this project since they're graduating. I want to help figure out how best to manage getting this merged in, since the CLAs show as needing their help but I can't rely on their availability since this is technically outside of their semester now.

Thanks,
Chris

**Comment by jimmy-long** [comment]
@googlebot I consent

**Comment by tonyniverth** [comment]
@googlebot I signed it!

**Comment by cdietschrun** [comment]
@mayurkale22 The authors that the bot needs are just different accounts for the same 2 people, Jimmy and Tony. Any chance you can help get the cla label set to yes?

**Comment by mayurkale22** [comment]
> @mayurkale22 The authors that the bot needs are just different accounts for the same 2 people, Jimmy and Tony. Any chance you can help get the cla label set to yes?

For now, I have manually verified the CLA.

**Comment by mayurkale22** [comment]
Please fix the build and unresolved comments.

File: packages/opencensus-exporter-azure/src/azure-stats.ts (hunk 2 above)
Comment (jimmy-long) [reply packages/opencensus-exporter-azure/src/azure-stats.ts:17]: Should be taken care of!

File: packages/opencensus-exporter-azure/src/index.ts (hunk 4 above)
Comment (jimmy-long) [reply packages/opencensus-exporter-azure/src/index.ts:18]: We were mimicking what we saw in the StackDriver exporter where both were included in the same package.

File: packages/opencensus-exporter-azure/src/azure-stats.ts (hunk 6 above)
Comment (jimmy-long) [reply packages/opencensus-exporter-azure/src/azure-stats.ts:54]: We created unit tests for both exporters that covered a lot of the functionality that was easily testable, and covered the rest with our blackbox tests located in the bbtest directory. Once you supply an Azure instrumentation key, you can easily run the tests and verify the functionality.

File: packages/opencensus-exporter-azure/src/azure-trace.ts (hunk 3 above)
Comment (jimmy-long) [reply packages/opencensus-exporter-azure/src/azure-trace.ts]: Done!

File: packages/opencensus-exporter-azure/package.json (hunk 1 above)
Comment (jimmy-long) [reply packages/opencensus-exporter-azure/package.json]: Fixed

**Comment by jimmy-long** [comment]
I went ahead and resolved comments that were a simple fix (adding license header, version bump, etc). Left two open that were related to design and testing. Happy to address any other concerns! Thanks.

**Comment by gdkerr** [comment]
@googlebot I signed it

**Comment by cdietschrun** [comment]
@mayurkale22 Thanks for the approve-The only CLA bump here is Gaven, but it's from an account I don't think he has access to (since it was with his university account and he's graduated). @gdkerr is Gaven Kerr's other account; I think you can probably just manually set the CLA as signed for him, or reach out to him to confirm.

**Comment by mayurkale22** [comment]
> @mayurkale22 Thanks for the approve-The only CLA bump here is Gaven, but it's from an account I don't think he has access to (since it was with his university account and he's graduated). @gdkerr is Gaven Kerr's other account; I think you can probably just manually set the CLA as signed for him, or reach out to him to confirm.

I have manually verified the CLA and merged the PR. Let me know if you are waiting for the release.

**Comment by cdietschrun** [comment]
Great news thanks! We'd like to pick up this release but it's not urgent (order of weeks would be our integration time)

**Comment by cdietschrun** [comment]
@mayurkale22 I'm just curious if you have a timeline for package release including this exporter? We have some resources that I would love to assign to integrate this coming soon.

**Comment by mayurkale22** [comment]
> @mayurkale22 I'm just curious if you have a timeline for package release including this exporter? We have some resources that I would love to assign to integrate this coming soon.

Apologies for the delay. I have cut the new release. See: https://www.npmjs.com/package/@opencensus/exporter-azure

Noticed two things:
1. README.md file is missing
2. I had to change publishConfig to cut the release. pls release (https://github.com/census-instrumentation/opencensus-node/pull/814)

**Comment by cdietschrun** [comment]
No need to apologize, this is great! Thank you!

**Pull request merged**